### PR TITLE
(#29) Fix version is not updated when parameter is changing

### LINKED ISSUE #29: Version is not updated when version parameter changes
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: all
- Ruby: all
- Distribution: all
- Module version: 3.0.0

## How to reproduce (e.g Puppet code you use)
Set $vault::install_method = 'archive'
Set $vault::version to e.g. '1.13.1'

## What are you seeing
The output of the command 'vault --version' is the old version i.e. 1.2.2

## What behaviour did you expect instead
I expect the module to update the executable i.e. output of the command 'vault --version' should be 1.13.1

### PR DESCRIPTION
**Pull request by C24-AK**
#### Pull Request (PR) description
This PR fixes the issue that version is not updated on parameter change.
This is due the nature of creates parameter in archive.

This PR is creating a fact of vaults version and compares it against the parameter.
If parameter is newer it will download the archive and extract it.

This will not reboot vault server.
Please be aware to restart vault server, so vault sources the new binary.

#### This Pull Request (PR) fixes the following issues
Fixes #29

### CODE REVIEW

**Comment by C24-AK** [comment]
Hey @bastelfreak ,are you able to review this PR? :)

File: lib/facter/vault_version.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+# Fact: vault_version
+#
+# Purpose: Retrieve vault version if installed
+#
+Facter.add(:vault_version) do
+  setcode do
+    if Facter::Util::Resolution.which('vault')
```
Comment (bastelfreak) [inline lib/facter/vault_version.rb:9]: I think the cleaner solution is to confine the fact 
```ruby
confine { Facter::Util::Resolution.which('ip') }
setcode do
 ...
```

Comment (C24-AK) [reply lib/facter/vault_version.rb]: Sorry for late reply. You are right, I changed the code accordingly.

**Comment by C24-AK** [comment]
Hey @bastelfreak, is there anything else needed? :) Looking forwarding for merging.

File: manifests/install.pp [hunk 2]
```diff
@@ -19,7 +19,13 @@
         extract_path => $vault::bin_dir,
         source       => $vault::real_download_url,
         cleanup      => true,
-        creates      => $vault_bin,
+        creates      => $facts['vault_version'] ? { # lint:ignore:selector_inside_resource
```
Comment (bastelfreak) [inline manifests/install.pp:22]: this while block is a bit hard to read, can you add a test for it?

Comment (C24-AK) [reply manifests/install.pp:22]: How would you do it?
I presume the fact will be nil on tests, since fact will be loaded before tests.

Comment (C24-AK) [reply manifests/install.pp:22]: I added the tests, could you check if they fit?

**Comment by C24-AK** [comment]
any updates @bastelfreak ?

File: spec/spec_helper_acceptance.rb [hunk 3]
```diff
@@ -5,6 +5,7 @@
 
 require 'voxpupuli/acceptance/spec_helper_acceptance'
 
+ENV['BEAKER_FACTER_VAULT_VERSION'] = '1.12.0'
```
Comment (bastelfreak) [inline spec/spec_helper_acceptance.rb:8]: That file is generated by a template. I took your changes, rebased them and adjusted it for the template: https://github.com/voxpupuli/puppet-vault/pull/55. Thanks a lot for your work!

Comment (C24-AK) [reply spec/spec_helper_acceptance.rb:8]: Looking forward for the changes :)

**Pull request merged**